src/batch.js

```diff
@@ -5,24 +5,228 @@ const { HttpError } = require('./httpError')
+  }
+
+  //Compose event to appear as native AWS event
+  let childEvent = {
+    ...event,
+    httpMethod: request.method,
+    headers,
+    body: request.body,
+    path: urlAndQueryString[0],
+    pathParameters: {
+      proxy: urlAndQueryString[0].substring(1)
+    },
+    multiValueQueryStringParameters: batch.composeQueryStringParametersFromUrl(request.url)
+  }
+
+  context._batch = true
+  context.response = undefined
+
+  return await route(childEvent, context)
+}
+
+function validateDependencyChain(requests) {
+  const dependencies = requests.reduce((obj, req) => {
+    obj[req.id] = { dependsOn: [], ...req }
+    return obj
+  }, {})
+  let keys = Object.keys(dependencies)
+  let length
+  let items
+
+  const sorted = new Set()
+  const result = []
+  do {
+    length = keys.length
+    items = []
+    // eslint-disable-next-line no-loop-func
+    keys = keys.filter(k => {
+      if (!dependencies[k].dependsOn.every(dependency => sorted.has(dependency))) {
+        return true
+      }
+      items.push(k)
+      return false
+    })
+    result.push(...items)
+    items.forEach(i => sorted.add(i))
+  } while (keys.length && keys.length !== length)
+
+  HttpError.assert(
+    keys.length === 0,
+    400,
+    `invalid body; unable to resolve execution order ` +
+      `due to circular dependency chain(s) involving: (${keys.join(', ')})`
+  )
+}
+
+function validateBatchRequest(body, maxBatchSize) {
+  // validate json body has been parsed
+  HttpError.assert(typeof body === 'object' && body !== null, 400, `invalid body; expected object`)
+  // validate body.requests has array of requests
+  HttpError.assert(
+    body.requests && Array.isArray(body.requests),
+    400,
+    `invalid body; expected requests to be an array`
+  )
+  HttpError.assert(
+    body.requests.length !== 0,
+    400,
+    `invalid body; requests must contain at least one element`
+  )
+  HttpError.assert(
+    body.requests.length <= maxBatchSize,
+    400,
+    `invalid body; max number of requests exceeded (${maxBatchSize})`
+  )
+
+  // validate each body.requests[]
+  const ids = body.requests.map(req => req.id)
+  for (let i = 0; i < body.requests.length; i++) {
+    const request = body.requests[i]
+    HttpError.assert(request.id, 400, `invalid body; requests[${i}].id required`)
+    HttpError.assert(
+      typeof request.id === 'string' || request.id instanceof String,
+      400,
+      `invalid body; requests[${i}].id must be a string`
+    )
+    HttpError.assert(request.method, 400, `invalid body; requests[${i}].method required`)
+    HttpError.assert(
+      ['GET', 'POST', 'PUT', 'PATCH', 'DELETE'].includes(request.method),
+      400,
+      `invalid body; requests[${i}].method must be one of ( GET | POST | PUT | PATCH | DELETE )`
+    )
+    HttpError.assert(request.url, 400, `invalid body; requests[${i}].url required`)
+    HttpError.assert(
+      typeof request.url === 'string' || request.url instanceof String,
```

Comment:
should we check for non empty strings?

Reply:
I believe that `request.url` check on L#202 accomplishes that.